srk/src/transition.ml

```diff
@@ -342,40 +342,87 @@ struct
+  (* helper method for interpolate/extrapolate procedures. creates fresh copies of skolem variables in tr *)
+  let rename_skolems tr =
+    let fresh_skolem =
+      Memo.memo (fun sym ->
+          match Var.of_symbol sym with
+          | Some _ -> mk_const srk sym
+          | None ->
+              let name = show_symbol srk sym in
+              let typ = typ_symbol srk sym in
+              mk_const srk (mk_symbol srk ~name typ))
     in
+    { transform = M.map (substitute_const srk fresh_skolem) tr.transform;
+      guard = substitute_const srk fresh_skolem tr.guard }
+
+
+
+  let interpolate_unsat_core solver trs post guards core = 
+    let refresh s = Smt.Solver.reset s; s 
+    in let core_symbols =
+      List.fold_left (fun core phi ->
+          match Formula.destruct srk phi with
+          | (`Proposition (`App (s, []))) -> Symbol.Set.add s core
+          | _ -> assert false)
+        Symbol.Set.empty
+        core
+    in
+    let (itp, _) =
+      List.fold_right2 (fun tr guard (itp, post) ->
+          let subst sym =
+            match Var.of_symbol sym with
+            | Some var ->
+              if M.mem var tr.transform then
+                M.find var tr.transform
+              else
+                mk_const srk sym
+            | None -> mk_const srk sym
+          in
+          let post' = substitute_const srk subst post in
+          let reduced_guard =
+            List.filter_map (fun (indicator, guard) ->
+                if Symbol.Set.mem indicator core_symbols then
+                  Some (mk_not srk guard)
+                else
+                  None)
+              guard
+          in
+          let wp =
+            (mk_not srk (mk_or srk (post'::reduced_guard)))
+            |> Quantifier.mbp srk ~solver:(refresh solver) (fun s -> Var.of_symbol s != None)
+            |> mk_not srk
+          in
+          (wp::itp, wp))
+        trs
+        guards
+        ([post], post)
+    in `Valid (List.tl itp)  
+
+
+  let interpolate_query ?(solver=Smt.mk_solver C.context) trs post sat_callback unsat_callback = 
+    let _ = Smt.Solver.reset solver in 
     (* Break guards into conjunctions, associate each conjunct with an indicator *)
     let guards =
       List.map (fun tr ->
           List.map
             (fun phi -> (mk_symbol srk `TyBool, phi))
             (destruct_and srk tr.guard))
-        trs
-    in
-    let indicators =
-      List.concat_map (List.map (fun (s, _) -> mk_const srk s)) guards
+        trs in 
+    let indicators, indicator_symbols =
+      List.concat_map (List.map (fun (s, _) -> mk_const srk s)) guards,
+      List.concat_map (List.map fst) guards |> Symbol.Set.of_list
     in
     let subscript_tbl = Hashtbl.create 991 in
+    let ss_inv = Hashtbl.create 991 in 
+    let sst = Hashtbl.create 991 in 
     let subscript sym =
       try
         Hashtbl.find subscript_tbl sym
       with Not_found -> mk_const srk sym
```

Comment:
To maintain the invariant that the domain of subscript_tbl is the set of all variables that are read/written, should insert sym -> (mk_const srk sym) when Var.of_symbol != None